tests/Add-DbaAgDatabase.Tests.ps1

```
@@ -38,49 +47,130 @@ Describe "Add-DbaAgDatabase" -Tag "UnitTests" {
+        # Use dbatools commands whenever possible, so that these are tested here as well.
+
+        # Explain what needs to be set up for the test:
+        # To add a database to an availablity group, we need an availability group and a database that has been backed up.
+        # For negative tests, we need a database without a backup and a non existing database.
+
+        # Set variables that we need in the tests.
+        # I don't see the point in prefixing them all with "dbatoolsci",
+        # better give them good names that relate to the command that we test.
+        $agName = "addagdb_group"
+        $existingDbWithBackup = "dbWithBackup"
+        $existingDbWithoutBackup = "dbWithoutBackup"
+        $nonexistingDb = "dbdoesnotexist"
+
+        # Create the objects
         $splatNewAg = @{
-            Primary = $TestConfig.instance3
-            Name = $agname
-            ClusterType = "None"
+            Primary      = $TestConfig.instance3
+            Name         = $agName
+            ClusterType  = "None"
             FailoverMode = "Manual"
-            Database = $dbname
-            Confirm = $false
-            Certificate = "dbatoolsci_AGCert"
+            Certificate  = "dbatoolsci_AGCert"
         }
-        $ag = New-DbaAvailabilityGroup @splatNewAg
+        $null = New-DbaAvailabilityGroup @splatNewAg
+
+        $null = New-DbaDatabase -SqlInstance $TestConfig.instance3 -Name $existingDbWithBackup
+        $backup = Backup-DbaDatabase -SqlInstance $TestConfig.instance3 -Database $existingDbWithBackup -Path $TestConfig.Temp
+        $filesToRemove += $backup.Path
+
+        $null = New-DbaDatabase -SqlInstance $TestConfig.instance3 -Name $existingDbWithoutBackup
+
+        # We want to run all commands outside of the BeforeAll block without EnableException to be able to test for specific warnings.
+        $PSDefaultParameterValues.Remove('*-Dba*:EnableException')
```

this, and the above, are going to be readily forgotten. I'm not sure this is a correct approach vs forcing -Ea Stop everywhere vs let the test fail at the first stages after beforeall. Of course, the one here is the sanest approach but making sure that everything is up and adherent to this template may be harder.
Maybe we should just set the runner preference to stop, AND to enable exception if not otherwise specified ?

---

I already have a pester test based on the AST that tests that - can be included like PSScriptAnalyser.

I forgot where it was, but I'm sure I have seen this: A test (incorrectly) passes because the setup failed silently. So for me it is super important that we check if the test setup is successful.

In addition I would from time to time test that there is no warning. If you look at the output of a AppVeyor run - there are a lot of warning that nobody cares about. I think that is super dangerous .